### PR TITLE
Handle env vars that use valueFrom

### PR DESCRIPTION
**Pull request by jhadvig**
When env vars have speciefied `valueFrom` instead of the `value` we need to render them as well in your `environments` directive.
Adding two proposed solutions that I've liked the most, but still not 100% sure about the solution. Think is that in case of the `configMapKeyRef` & `secretKeyRef` there could be more then 'key' and 'name' based on the https://github.com/kubernetes/kubernetes/blob/b45bf88105159cdbdd112e899bcb073e2b1ae573/pkg/api/types.go#L2048-L2050 so I didnt wanted to have a separate column for each field. On the other hand I wanted to stay by the table concept since the default original envVars are using it.
screenshot-11
screenshot-12
The code is not in the the best shape so for now please skip the code review.

@jwforres @spadgett @ajacobs21e thoughts?

closes https://github.com/openshift/origin/issues/7886

### CODE REVIEW

**Comment by spadgett** [comment]
If we don't have a consistent set of columns we can use, I don't think a table is right. Maybe just have a referenced variables section with definition lists for each variable?

**Comment by spadgett** [comment]
Yeah I was thinking a "Dynamic Variables" section or similar. (I'm OK with a different name there.)

Forgive the ASCII art.

```
 -------------
| Environment |

 -------------
| Name | Val  |
| FOO  | bar  |
| ...  |      |
 -------------

Dynamic Variables

POD_NAMESPACE
Type: ...
Field Path: ...

MY_SECRET_DATA
...
```

**Comment by jhadvig** [comment]
@spadgett I've addressed the comments. PTAL
screenshot-13

**Comment by jwforres** [comment]
we've been moving towards humanizing types everywhere, maybe Field instead
of fieldRef and Secret Key instead of secretKeyRef.  you could do it
generically (so we handle future types) by replacing "Ref" with "" and then
using @spadgett's helper that does the splitting / upper-casing

**Comment by jwforres** [comment]
I'm wondering if the Container header needs to be bigger, right now it
looks like the Container and the Referenced Variables headers are the same
size, when Ref Vars should be a sub-section of the container.

**Comment by jwforres** [comment]
I can't decide if I like the the hr between the different referenced
variables, i think that's inconsistent with what we do for volumes, volumes
are very similar, they have types and each type has different metadata
associated with it.  I know @rhamilto has been looking at that Volumes
section some so CC'ing here

**Comment by jhadvig** [comment]
@jwforres +1 for making the container header bigger

**Comment by jhadvig** [comment]
Regarding the delimiter I wasnt sure as well, but put it there anyway so I get some feedback. In fact I kinda like this a bit more since you can optically differentiate between the envVars. But yeah otherwise its similar with the volumes

**Comment by jwforres** [comment]
So looking at this for a running container... I would really like to know what value that environment reference was resolved to.  AFAIK not possible with the current api response.  @smarterclayton @deads2k is that even technically possible to do later?

**Comment by jhadvig** [comment]
Yeah I dont think that is possielbe ATM

**Comment by rhamilto** [comment]
> I can't decide if I like the the hr between the different referenced
variables, i think that's inconsistent with what we do for volumes, volumes
are very similar, they have types and each type has different metadata
associated with it.  I know @rhamilto has been looking at that Volumes
section some so CC'ing here

The HR is inconsistent with what we're doing on Volumes [1], and I feel the HR is unnecessary noise.  

[1] https://github.com/openshift/origin/pull/7937.

**Comment by jhadvig** [comment]
@rhamilto went throught your PR and I like the styling. Will update accordingly :+1:

**Comment by deads2k** [comment]
> So looking at this for a running container... I would really like to know what value that environment reference was resolved to. AFAIK not possible with the current api response. @smarterclayton @deads2k is that even technically possible to do later?

You mean, "can you determine the value you got for `valueFrom.fieldRef.fieldPath`" sort of thing?  Not in a way that will make you happy.  I think only the kubelet and docker know the value that it used and we don't have an API for "get me metadata for this actual container".

You could write a parser that takes a good guess assuming stable-ish values for things you're referencing, but if they've changed that's not necessarily what the kubelet used.

**Comment by jwforres** [comment]
On Fri, Mar 11, 2016 at 9:15 AM, David Eads <notifications@github.com>
wrote:

> So looking at this for a running container... I would really like to know
> what value that environment reference was resolved to. AFAIK not possible
> with the current api response. @smarterclayton
> <https://github.com/smarterclayton> @deads2k <https://github.com/deads2k>
> is that even technically possible to do later?
>
> You mean, "can you determine the value you got for
> valueFrom.fieldRef.fieldPath" sort of thing? Not in a way that will make
> you happy. I think only the kubelet and docker know the value that it used
> and we don't have an API for "get me metadata for this actual container".
>

Yeah "get me metadata for this actual container" is really what I was
hoping for, that's a shame :-\

so @jhadvig maybe we have a tip that says something like "( i ) On running
pods you can check what these reference values have resolved to by going to
the Terminal tab and echoing the environment variable."    @spadgett
thoughts?

**Comment by jhadvig** [comment]
@jwforres @rhamilto I've addressed your comments. PTAL
screenshot-14
screenshot-15

**Comment by smarterclayton** [comment]
I would expect the layout of these to be in the container table.  This
makes them look special.  Also, it ignores ordering, and ordering is
critical.  The order of the env vars in the pod should *never* be changed
on display.

**Comment by smarterclayton** [comment]
Because you can turn around and do env(10) FOO=$(MY_SECRET_DATA) and that
is a different outcome when env(10) is before or after MY_SECRET_DATA's
definition.

**Comment by jwforres** [comment]
@jhadvig given @smarterclayton's feedback we have to put it back into the table and we need to double check that the table ordering is exactly what we are getting back from the API.

so to distinguish these values in the table as being "special" I was going to propose putting http://fontawesome.io/icon/external-link-square/ in front of the value when they are references

and instead of calling out the various field names do something like

[->] Secret foo, key bar
[->] Field metadata.namespace
etc...

**Comment by jhadvig** [comment]
@jwforres I've updated the PR based on our and @smarterclayton comments. Hopefully my styling estimations were accurate. PTAL
screenshot-19
screenshot-20

**Comment by jwforres** [comment]
I was thinking more like a one line summary of the from information, and don't duplicate the environment name, so for your two examples it would be

```
|  POD_NAMESPACE   |  [->] Field metadata.namespace        |
|  MY_SECRET_DATA  |  [->] Secret test.secret, key data-1  |
```

where | is my ascii art to indicate the columns

**Comment by jhadvig** [comment]
@jwforres updated
screenshot-23

**Comment by smarterclayton** [comment]
The value column is hard to read, i have no idea what is a user entered
value vs descriptive text.

**Comment by jhadvig** [comment]
@smarterclayton had it like this:

screenshot-22

But we decided with @jwforres and @spadgett to dont distinguish between key and value.

**Comment by smarterclayton** [comment]
It's really, really, really hard to parse when it's just text.  This is a
structured field, it should field like it in some form.  I'm not even sure
the current text adequately conveys what each one does.

**Comment by jhadvig** [comment]
@smarterclayton @jwforres updated. As suggested I've put there a sentence that would make thing more clear, together with the popover + the sentence uses different font
screenshot-25

File: assets/app/views/directives/environment.html [hunk 1]
```diff
@@ -1,10 +1,33 @@
...
+              Set to the key <b>{{env.ref.key}}</b> in secret <b>{{env.ref.name}}.</b>
+            </span>
+            <span ng-switch-when="secretKeyRef">
+              Set to the key <b>{{env.ref.key}}</b> in secret <b>{{env.ref.name}}.</b>
+            </span>
+            <span ng-switch-when="fieldRef">
+              Set to the field <b>{{env.ref.fieldPath}}</b> in the current object.
+            </span>
+            <span ng-switch-default>
+              {{env.type}}
```
Comment (jwforres) [inline assets/app/views/directives/environment.html]: Maybe `Set to a reference on a <b>{{env.type}}</b>`

File: assets/app/scripts/directives/resources.js [hunk 2]
```diff
@@ -145,6 +145,24 @@ angular.module('openshiftConsole')
       templateUrl: 'views/directives/environment.html',
       controller: function($scope) {
         $scope.expanded = {};
+        var newEnvVars = [];
+
+        angular.forEach($scope.envVars, function(envVar) {
```
Comment (jwforres) [inline assets/app/scripts/directives/resources.js]: is this normalization necessary? you are using a switch to handle different cases already in the view.

if you keep it you have to add a link function with a watch for envVars, controller does not get re-run if the data in the binding changes.

File: assets/app/styles/_core.less [hunk 3]
```diff
@@ -976,6 +976,16 @@ h1 small.meta, .build-config-summary .meta {
...
     }
   }
+  .action-inline {
+    margin-left: 0;
+    font-size: @font-size-base;
+    i.fa {
+      color: @btn-default-color;
+    }
+  }
+  .dl-horizontal {
```
Comment (jwforres) [inline assets/app/styles/_core.less]: shouldnt be needed anymore since you dropped the dl

File: assets/app/styles/_core.less [hunk 4]
```diff
@@ -976,6 +976,16 @@ h1 small.meta, .build-config-summary .meta {
       }
     }
   }
+  .action-inline {
```
Comment (jwforres) [inline assets/app/styles/_core.less]: i don't see anything with the action-inline class being added, is this still needed?

**Comment by jhadvig** [comment]
@jwforres I've reverted the normalization, you were right it wants necessary. PTAL

**Comment by jwforres** [comment]
:+1:  LGTM

**Comment by jwforres** [comment]
[merge]
